jicofo-selector/src/main/kotlin/org/jitsi/jicofo/bridge/Cascade.kt

```diff
@@ -0,0 +1,330 @@
+        }
+        val other = sessions[key]
+        checkNotNull(other) {
+            "$node has link to $key not found in cascade"
+        }
+        val backLink = other.relays[node.relayId]
+        checkNotNull(backLink) {
+            "$node has link to $other, but $other has no link to $node"
+        }
+        check(link.meshId == backLink.meshId) {
+            "$node links to $other in mesh ${link.meshId}, but backlink has mesh ${backLink.meshId}"
+        }
+    }
+}
+
+/** Validate a mesh, or throw IllegalStateException. */
+fun <N : CascadeNode<N, L>, L : CascadeLink> Cascade<N, L>.validateMesh(meshId: String?) {
+    val meshNodes = getMeshNodes(meshId)
+
+    meshNodes.forEach { node ->
+        meshNodes.forEach { other ->
+            if (other != node) {
+                check(node.relays.contains(other.relayId)) {
+                    "$node has no link to $other in mesh $meshId"
+                }
+                val link = node.relays[other.relayId]
+                check(link!!.meshId == meshId) {
+                    "link from $node to $other has meshId ${link.meshId}, not expected $meshId"
+                }
+            }
+        }
+    }
+}
+
+private fun <N : CascadeNode<N, L>, L : CascadeLink> Cascade<N, L>.visitNodeForValidation(
+    node: N,
+    parent: N?,
+    visitedNodes: MutableSet<String?>,
+    validatedMeshes: MutableSet<String?>
+) {
+    validateNode(node)
+    visitedNodes.add(node.relayId)
+
+    node.relays.values.forEach {
+        if (!visitedNodes.contains(it.relayId)) {
+            if (!validatedMeshes.contains(it.meshId)) {
+                validateMesh(it.meshId)
+                validatedMeshes.add(it.meshId)
+            }
+            val linkedNode = sessions[it.relayId]
+            checkNotNull(linkedNode) {
+                "$node has link to node ${it.relayId} not found in cascade"
+            }
+            visitNodeForValidation(linkedNode, node, visitedNodes, validatedMeshes)
+        } else {
+            check(it.relayId == parent?.relayId || validatedMeshes.contains(it.meshId)) {
+                "Multiple paths found to ${sessions[it.relayId]}"
+            }
+        }
+    }
+}
+
+/** Validate a cascade, or throw IllegalStateException */
+fun <N : CascadeNode<N, L>, L : CascadeLink> Cascade<N, L>.validate() {
+    if (sessions.isEmpty()) {
+        /* Empty cascade is trivially valid */
+        return
+    }
+    val firstNode = sessions.values.first()
+
+    val visitedNodes = HashSet<String?>()
+    val validatedMeshes = HashSet<String?>()
+
+    visitNodeForValidation(firstNode, null, visitedNodes, validatedMeshes)
+
+    check(visitedNodes.size == sessions.size) {
+        val unvisitedNodes = sessions.keys.subtract(visitedNodes)
+        "Nodes ${unvisitedNodes.joinToString()} not reachable from initial node $firstNode"
+    }
+}
+
+/* Get the distance, in hops in the cascade, from one node to a node satisfying a property;
```

Comment:
Is this supposed to find the *shortest* distance? I guess the way you use it it doesn't matter, but with some predicates the distance would be different depending on how the relays are traversed.